cpp/farm_ng/core/misc/shared.h

```diff
@@ -0,0 +1,191 @@
+  T* operator->() { return non_null_shared_.get(); }
+
+  /// Returns the interior object which is guarenteed to be available
+  T const* operator->() const { return non_null_shared_.get(); }
+
+  // Implicit conversion to a nullable std::shared_ptr okay
+  operator std::shared_ptr<T>() const { return sharedPtr(); }
+
+  // Return a nullable shared_ptr<T> from this Shared<T> object
+  std::shared_ptr<T> sharedPtr() const { return non_null_shared_; }
+
+  // Return the raw point from this Shared<T> object
+  T const* ptr() const { return sharedPtr().get(); }
+
+  // Return the raw point from this Shared<T> object
+  T* ptr() { return sharedPtr().get(); }
+
+  // Default constructable only if T is
+  Shared() requires(std::is_default_constructible<T>::value)
+      : non_null_shared_(std::make_shared<T>()) {
+    checkMaybeThrow();
+  }
+
+  // Copy constructor from derived bases
+  template <DerivedFrom<T> Derived>
+  Shared(Shared<Derived> const& other) : non_null_shared_(other.sharedPtr()) {
+    checkMaybeThrow();
+  }
+
+  // Construct from shared_ptr
+  template <DerivedFrom<T> Derived>
+  Shared(std::shared_ptr<Derived> const& panic_if_null)
+      : non_null_shared_(panic_if_null) {
+    checkMaybeThrow();
+  }
+
+  // Take ownership from unique_ptr
+  template <DerivedFrom<T> Derived>
+  Shared(std::unique_ptr<Derived>&& panic_if_null)
+      : non_null_shared_(std::move(panic_if_null)) {
+    checkMaybeThrow();
+  }
+
+  // Not sure why this is needed when the generic one
+  // is defined above
+  Shared(std::shared_ptr<T> const& panic_if_null)
+      : non_null_shared_(panic_if_null) {
+    checkMaybeThrow();
+  }
+
+  Shared(Shared<T>&& o) : non_null_shared_(o.non_null_shared_) {
+    // We mustn't move the internal shared_ptr
+    // because that would break the invariant.
+  }
+
+  Shared<T>& operator=(Shared<T>&& o) {
+    // We maintain the invariant since
+    // o.non_null_shared_ must also be valid
+    non_null_shared_ = o.non_null_shared_;
+    return *this;
+  }
+
+  Shared(Shared<T> const&) = default;
+  Shared<T>& operator=(Shared<T> const&) = default;
+
+  bool operator==(Shared<T> const& rhs) const noexcept {
+    return this->non_null_shared_ == rhs.non_null_shared_;
+  }
+
+ private:
+  void checkMaybeThrow() const {
+    if (!non_null_shared_) {
+      throw BadExpectedAccess(FARM_ERROR_REPORT("Shared is null."));
+    }
+  }
+
+  // Class invariant:non_null_shared_ is guaranteed not to be null.
+  std::shared_ptr<T> non_null_shared_;
+};
+
+template <typename T>
+using ExpectShared = farm_ng::Expected<Shared<T>>;
+
+template <class T, class... Args>
+static ExpectShared<T> tryMakeShared(Args&&... args) noexcept {
+  return Shared<T>::tryMake(std::forward<Args>(args)...);
+}
+
+/// + represents a checked dereference similar to operator*()
+/// Unlike Expected::operator*() which will allow for UB,
+/// + will Panic if x is in the error state.
+template <class T>
+Shared<T>& operator+(ExpectShared<T>& x) {
```

Comment:
I think we should scrap these operator+ methods. I haven't been using them as I think they're a bit confusing.